README
Nagios::NRPE 
============

An implementation of NRPE in pure perl

INSTALLATION
------------

To install it: 

   perl Build.PL
   ./Build
   ./Build test 
   ./Build install

SCRIPTS
-------

Inside bin/ you'll find an implementation of the check_nrpe binary normally
shipped with Nagios' nrpe server.

DOCUMENTATION
-------------

Every module shipped with this package has a full documentation in pod format. 
To read it use perldoc like this:
   
   perldoc Nagios::NRPE::Client

In the case of problems or appreciation please visit these URLs

*   [Github issues](https://github.com/stockholmuniversity/nagios-nrpe/issues) 

*   [CPAN Ratings](http://cpanratings.perl.org/d/Nagios-NRPE)

*   [Search CPAN](http://search.cpan.org/dist/Nagios-NRPE)

*   [Github repository](https://github.com/stockholmuniversity/nagios-nrpe/)

COPYRIGHT AND LICENSE
---------------------

This software is copyright (c) 2013-2018 by the authors (see AUTHORS file).

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
2.0.13 2020-07-10
 - No changes - just META information for CPAN

2.0.12 2020-07-10
 - No changes - just META information for CPAN

2.0.11 2020-06-12
 - Use default SSL_version from IO::Socket::SSL

2.0.10 2019-06-24
 - Error messages are always returned has hashes.

2.0.9 2018-12-03
 - Fix typo in manifest

2.0.8 2018-11-29
 - Yet another attempt to fix CPANTS warnings for manifest

2.0.7 2018-11-29
 - Another attempt to fix CPANTS warnings for manifest

2.0.6 2018-11-28
 - Fix manifest and meta

2.0.5 2018-11-28
 - Move podcover test to subfolder
 - Add timeout for SSL/TLS connections to stop possible DOS-attack

2.0.4 2018-11-23
 - Don't run covertest on end users
 - Handle servers that respond with a v2 packet to a v3 request (NSClient++)

2.0.3 2018-11-21
 - Don't require Data::Dumper - trying to make CPAN Tester happy

2.0.2 2018-11-15
 - Fix crc32 and padding problem

2.0.1 2018-11-09
 - Create valid v3 Packets.
 - Don't crash on invalid Packets.
 - Don't crash on with Packets from an unauthorized IP.

2.0.0 2018-09-20
 - Fix missing exit codes from check
 - NRPE v3 now working properly with "short" outputs
 - Better SSL handling
 - Deassemble function removed

1.0.3 2017-08-22
 - Add Utils.pm that contains common helper functions.
 - Fix error handling in Packet.pm
 - Deprecate deassemble function
 - Add disassemble function

1.0.2 2017-07-27
 - Fix manifest.

1.0.1 2017-07-27
 - Add support for NRPE V3 and IPv6.

1.0.0 2016-09-12
 - Sort options alphabetically

0.004 2016-06-01
 - Fix SSL incompatibility with recent NRPE Server implementations

0.003 2014-04-16
 - Fix Pod errors

0.002 2014-04-15
 - Fix kwalitee errors
 - Build.PL add Test::Podcoverage to build requirements
 - add license to all *.pl and *.pm files

0.001 2013-01-10
 - First release.


Build.PL
#!/usr/bin/perl
use 5.010_000;

use strict;
use warnings;
use Module::Build;

my $build = Module::Build->new(
    module_name => 'Nagios::NRPE',
    license     => 'perl',
    dist_author => [
        'Andreas Marschke',
        'Sven Nierlein',
        'Franz Skale',
        'Johan Wassberg <jocar@su.se>',
        'Christian Biamont <cbeam@su.se>',
        'Micke Nordin <mikael.nordin@su.se>',
        'Peter Valdemar Mørch'
    ],
    meta_merge => {
        resources => {
            homepage   => 'http://github.com/stockholmuniversity/nagios-nrpe',
            repository => 'git://github.com/stockholmuniversity/nagios-nrpe.git'
        },
        no_index => {
            directory => [qw/t xt/],
        }
    },
    configure_requires => {
        'Module::Build' => '0.31',
    },
    build_requires => {
        "Test::More"          => 0,
        "Test::Pod::Coverage" => 0
    },
    requires => {
        "perl"               => "5.010_000",
        "Convert::Binary::C" => 0,
        "constant"           => 0,
        "Digest::CRC"        => 0,
        "IO::Socket"         => 0,
        "IO::Socket::INET6"  => 0,
        "Net::Subnet"        => 0,
        "Time::HiRes"        => 0
    },
    dynamic_config       => 0,
    recursive_test_files => 1,
    create_license       => 1,
    test_files           => 't/*.t xt/*.t',
);

$build->create_build_script;


lib/Nagios/NRPE.pm
=head1 NAME

Nagios::NRPE - A Nagios NRPE implementation in pure perl

=head1 SYNOPSIS

 # Executing a check on an NRPE-Server
 use Nagios::NRPE::Client;

 my $client = Nagios::NRPE::Client->new( host => "localhost", check => 'check_cpu');
 my $response = $client->run();
 if(defined $response->{error}) {
   print "ERROR: Couldn't run check ".$client->check()." because of: ".$response->{reason}."\n";
 } else {
   print $response->{status}."\n";
 }

 # Reading and Writing Nagios NRPE Packets

 use IO::Socket;
 use IO::Socket::INET;
 # Import necessary constants into Namespace
 use Nagios::NRPE::Packet qw(NRPE_PACKET_VERSION_3
                             NRPE_PACKET_QUERY
                             STATE_UNKNOWN
                             STATE_CRITICAL
                             STATE_WARNING
                             STATE_OK);

 my $packet = Nagios::NRPE::Packet->new();

 my $socket = IO::Socket::INET->new(
                    PeerAddr => $host,
                    PeerPort => $port,
                    Proto    => 'tcp',
                    Type     => SOCK_STREAM) or die "ERROR: $@ \n";

 print $socket $packet->assemble(type => NRPE_PACKET_QUERY,
                              buffer => "check_load 1 2 3",
                              version => NRPE_PACKET_VERSION_3 );

 my $data = <$socket>;
 my $response = $packet->disassemble($data);

 print $response->{buffer};

=head1 DESCRIPTION

This file currently only serves as a stub so Build.PL will find it. For more information on
the submodules please read L<Nagios::NRPE::Client> or L<Nagios::NRPE::Packet> or L<Nagios::NRPE::Daemon>.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013-2018 by the authors (see L<AUTHORS|https://github.com/stockholmuniversity/Nagios-NRPE/blob/master/AUTHORS> file).

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

package Nagios::NRPE;

use strict;
use warnings;

our $VERSION = '2.0.13';

1;


share/nrpe-testing.pl
#!/usr/bin/perl
use 5.010_000;

use strict;
use warnings;

use Data::Dumper;
use IO::Socket;
use IO::Socket::INET;
use constant {
  # packet version identifier
  NRPE_PACKET_VERSION_3   =>  3,
  NRPE_PACKET_VERSION_2   =>  2,
  NRPE_PACKET_VERSION_1   =>  1,

  # id code for queries and responses to queries
  QUERY_PACKET            =>  1,
  RESPONSE_PACKET         =>  2,
  # max amount of data we'll send in one query/response
  MAX_PACKETBUFFER_LENGTH => 1024,

  MAX_COMMAND_ARGUMENTS   => 16,
  NRPE_HELLO_COMMAND      => "_NRPE_CHECK",
  DEFAULT_SOCKET_TIMEOUT  => 10,
  DEFAULT_CONNECTION_TIMEOUT => 300,

  # /* service state return codes */
  STATE_UNKNOWN           => 3,
  STATE_CRITICAL          => 2,
  STATE_WARNING 	  => 1,
  STATE_OK                => 0
};
my ($packet_version,  # Version of NRPE
    $packet_type,     # Clients = QUERY = 1, Server = RESPONSE = 2
    $crc32_value,
    $result_code,     # \x00 on querys
    $buffer           # buffer text consisting of the check name and its adjacent arguments max 1024 bytes
);

my $command = "check_users";

# Packet data
$packet_version = pack('n',NRPE_PACKET_VERSION_2);
$packet_type = pack('n',QUERY_PACKET);
$result_code = pack('n',2324);
$buffer = pack('a[1024]',$command);

my $packet = $packet_version.$packet_type."\x00\x00\x00\x00".$result_code.$buffer."\x00\x00";

$crc32_value = ~_crc32($packet);
my $packed_crc = pack ( "N", $crc32_value);

$packet = $packet_version.$packet_type.$packed_crc.$result_code.$buffer."\x00\x00";

my $socket = IO::Socket::INET->new(PeerAddr => '10.0.0.47',
				   PeerPort => '5666',
				   Proto => 'tcp',
				   Type => SOCK_STREAM) or die "ERROR: $@ \n";

print $socket $packet;

while (<$socket>) {
  my ($packet_version,$packet_type,$crc32,$result_code,$buffer) = unpack('nnNna*',$_);
  print "Packet Version: $packet_version\n";
  print "Packet Type: $packet_type\n";
  print "CRC32 Checksum: $crc32\n";
  print "Result_code: $result_code\n";
  print "Buffer: $buffer\n";
}

close($socket);

=pod

Utils Section

=cut

sub packet_dump {
  my $packet = shift;
  my ($i, $k, $dump, $l, $ascii, $c);

  for ( $i = 0; $i < length ( $packet ); $i+=16 ) {
    $l = $i+16;
    if ( $l > length ( $packet) ) {
      $l = length($packet);
    }
    $dump   = sprintf ( "%04d - %04d: ", $i, $l );
    $ascii  = "";
    for ( $k = $i; $k < $l; $k++ ) {
      $c     = ( ord ( substr ( $packet, $k, 1 ) ) );
      $dump .= sprintf ( "%02x ", $c );
      if (( $c >= 32 ) && ( $c <= 126 )) {
	$ascii .= chr ( $c );
      } else {
	$ascii .= ".";
      }
    }
    for ( $k = 0; $k < ( $i + 16 - $l ); $k++ ) {
      $dump .= "   ";
    }
    print( "packet_dump() ".$dump." [".$ascii."]\n" );
  }
}

sub _crc32 {
  my $crc;
  my $len;
  my $i;
  my $index;
  my @args;
  my ($arg) = @_;
  my @crc_table =(
    0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419,
    0x706af48f, 0xe963a535, 0x9e6495a3, 0x0edb8832, 0x79dcb8a4,
    0xe0d5e91e, 0x97d2d988, 0x09b64c2b, 0x7eb17cbd, 0xe7b82d07,
    0x90bf1d91, 0x1db71064, 0x6ab020f2, 0xf3b97148, 0x84be41de,
    0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7, 0x136c9856,
    0x646ba8c0, 0xfd62f97a, 0x8a65c9ec, 0x14015c4f, 0x63066cd9,
    0xfa0f3d63, 0x8d080df5, 0x3b6e20c8, 0x4c69105e, 0xd56041e4,
    0xa2677172, 0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b,
    0x35b5a8fa, 0x42b2986c, 0xdbbbc9d6, 0xacbcf940, 0x32d86ce3,
    0x45df5c75, 0xdcd60dcf, 0xabd13d59, 0x26d930ac, 0x51de003a,
    0xc8d75180, 0xbfd06116, 0x21b4f4b5, 0x56b3c423, 0xcfba9599,
    0xb8bda50f, 0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924,
    0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d, 0x76dc4190,
    0x01db7106, 0x98d220bc, 0xefd5102a, 0x71b18589, 0x06b6b51f,
    0x9fbfe4a5, 0xe8b8d433, 0x7807c9a2, 0x0f00f934, 0x9609a88e,
    0xe10e9818, 0x7f6a0dbb, 0x086d3d2d, 0x91646c97, 0xe6635c01,
    0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e, 0x6c0695ed,
    0x1b01a57b, 0x8208f4c1, 0xf50fc457, 0x65b0d9c6, 0x12b7e950,
    0x8bbeb8ea, 0xfcb9887c, 0x62dd1ddf, 0x15da2d49, 0x8cd37cf3,
    0xfbd44c65, 0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2,
    0x4adfa541, 0x3dd895d7, 0xa4d1c46d, 0xd3d6f4fb, 0x4369e96a,
    0x346ed9fc, 0xad678846, 0xda60b8d0, 0x44042d73, 0x33031de5,
    0xaa0a4c5f, 0xdd0d7cc9, 0x5005713c, 0x270241aa, 0xbe0b1010,
    0xc90c2086, 0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
    0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4, 0x59b33d17,
    0x2eb40d81, 0xb7bd5c3b, 0xc0ba6cad, 0xedb88320, 0x9abfb3b6,
    0x03b6e20c, 0x74b1d29a, 0xead54739, 0x9dd277af, 0x04db2615,
    0x73dc1683, 0xe3630b12, 0x94643b84, 0x0d6d6a3e, 0x7a6a5aa8,
    0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1, 0xf00f9344,
    0x8708a3d2, 0x1e01f268, 0x6906c2fe, 0xf762575d, 0x806567cb,
    0x196c3671, 0x6e6b06e7, 0xfed41b76, 0x89d32be0, 0x10da7a5a,
    0x67dd4acc, 0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5,
    0xd6d6a3e8, 0xa1d1937e, 0x38d8c2c4, 0x4fdff252, 0xd1bb67f1,
    0xa6bc5767, 0x3fb506dd, 0x48b2364b, 0xd80d2bda, 0xaf0a1b4c,
    0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55, 0x316e8eef,
    0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236,
    0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f, 0xc5ba3bbe,
    0xb2bd0b28, 0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7, 0xb5d0cf31,
    0x2cd99e8b, 0x5bdeae1d, 0x9b64c2b0, 0xec63f226, 0x756aa39c,
    0x026d930a, 0x9c0906a9, 0xeb0e363f, 0x72076785, 0x05005713,
    0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38, 0x92d28e9b,
    0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242,
    0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1,
    0x18b74777, 0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c,
    0x8f659eff, 0xf862ae69, 0x616bffd3, 0x166ccf45, 0xa00ae278,
    0xd70dd2ee, 0x4e048354, 0x3903b3c2, 0xa7672661, 0xd06016f7,
    0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc, 0x40df0b66,
    0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
    0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605,
    0xcdd70693, 0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8,
    0x5d681b02, 0x2a6f2b94, 0xb40bbe37, 0xc30c8ea1, 0x5a05df1b,
    0x2d02ef8d
   );

  $crc = 0xffffffff;
  $len = length($arg);
  @args = unpack "c*", $arg;
  for ($i = 0; $i < $len; $i++) {
    $index = ($crc ^ $args[$i]) & 0xff;
    $crc = $crc_table[$index] ^ (($crc >> 8) & 0x00ffffff);
  }
  return $crc;
}



lib/Nagios/NRPE/Utils.pm
=head1 NAME

Nagios::NRPE::Utils - Common helper functions for Nagios::NRPE

=head1 DESCRIPTION

This file contains common helper functions for the submodules 
Nagios::NRPE::Client,  Nagios::NRPE::Packet and Nagios::NRPE::Daemon.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013-2018 by the authors (see L<AUTHORS|https://github.com/stockholmuniversity/Nagios-NRPE/blob/master/AUTHORS> file).
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 METHODS

=over  

=item return_error

Create a hash with the specified error message using this format

{
    error => 1,
    reason => "some reason"
}

=back

=cut

package Nagios::NRPE::Utils;

our $VERSION = '2.0.13';
use strict;
use warnings;
require Exporter;

our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(return_error);

sub return_error {
    my ($reason) = @_;
    my %return;
    $return{'error'}  = 1;
    $return{'reason'} = $reason;
    return ( \%return );

}

1;


lib/Nagios/NRPE/Packet.pm
=head1 NAME

Nagios::NRPE::Packet - Assembly and disassembly of an NRPE packet

=head1 SYNOPSIS

 use IO::Socket;
 use IO::Socket::INET;
 # Import necessary constants into Namespace
 use Nagios::NRPE::Packet qw(NRPE_PACKET_VERSION_3
                             NRPE_PACKET_QUERY
                             MAX_PACKETBUFFER_LENGTH
                             STATE_UNKNOWN
                             STATE_CRITICAL
                             STATE_WARNING
                             STATE_OK);

 my $packet = Nagios::NRPE::Packet->new();

 my $socket = IO::Socket::INET->new(
                    PeerAddr => $host,
                    PeerPort => $port,
                    Proto    => 'tcp',
                    Type     => SOCK_STREAM) or die "ERROR: $@ \n";

 print $socket $packet->assemble(type => NRPE_PACKET_QUERY,
                              buffer => "check_load 1 2 3",
                              version => NRPE_PACKET_VERSION_3 );

 my $data = <$socket>;
 my $response = $packet->disassemble($data);

 print $response->{buffer};
 exit $response->{result_code};

=head1 DESCRIPTION

This class is meant to be used when an active connection exists and is ready to send the
packet.

=head1 CONSTRUCTION

=over

=item new

Takes the following options as a hashref

=back

=head1 SUBROUTINES

Following functions can be used after the creation of the packet

=over 2

=item assemble()

Takes a hash of options defining the packet to be sent and returns the assembled packet. You can print this
to an open socket and send it to either a server or the client depending on your situation.

 * check

A string defining the check to be run or the output of a check eg: "check_cpu"
NOTE: Nagios can accept arguments appended to the check in the form: "check_somecheck!ARG1!ARG2!ARG..."

 * version

The NRPE version you want to use (only V2 and V3 work V1 is not supported, deafult is V3).

See CONSTANTS for options here.

 * type

The TYPE of packet you wish to send, which is either QUERY or RESPONSE.

See L</CONSTANTS> for options here.

 * result_code

This is the exit code of the check script that is run, and check_nrpe.pl will exit with this value from the 
RESPONSE packet.

A set value for the QUERY type packet is 2324.

=item assemble_v2()

A helper function to assemble a V2 packet.

=item assemble_v3()

A helper function to assemble a V3 packet.

=item disassemble()

Takes a packet recieved by either client or server and disassembles them. The returned hashref contains 
the following values for a V3 packet:

 packet_version 
 packet_type    
 crc32_value    
 result_code    
 alignment      
 buffer_length  
 buffer

and the following values for a V2 packet:

 packet_version 
 packet_type    
 crc32_value    
 result_code    
 buffer

=item disassemble_v2()

Helper function for disassembleing a V2 packet

=item disassemble_v3()

Helper function for disassembleing a V3 packet

=item validate($packet)

Validates the contents of a packet using CRC32 checksumming. Returns undef
if not succesful.


=item packet_dump

Debugging function for hexdumping a binary string.

=back

=head1 CONSTANTS

These constants can be exported upon request with the 'use' pragma like this:

 # Will only import the constant NRPE_PACKET_VERSION_3 into your namespace
 use Nagios::NRPE::Packet qw(NRPE_PACKET_VERSION_3);

=over 2

=item * NRPE_PACKET_VERSION_3
        NRPE_PACKET_VERSION_2
        NRPE_PACKET_VERSION_1

The value of the NRPE version you want/need to use.

=item * NRPE_PACKET_QUERY
        NRPE_PACKET_RESPONSE

The packet type you want to send or recieve

=item * MAX_PACKETBUFFER_LENGTH
        MAX_COMMAND_ARGUMENTS

A threshhold on the send data

=item * NRPE_HELLO_COMMAND

unknown

=item * DEFAULT_SOCKET_TIMEOUT
        DEFAULT_CONNECTION_TIMEOUT

The default timeout for a connection and its corresponding socket

=item * STATE_UNKNOWN
        STATE_CRITICAL
        STATE_WARNING
        STATE_OK

States returned by the check

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013-2018 by the authors (see L<AUTHORS|https://github.com/stockholmuniversity/Nagios-NRPE/blob/master/AUTHORS> file).

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

package Nagios::NRPE::Packet;

our $VERSION = '2.0.13';

use 5.010_000;
require Exporter;
require overload;

BEGIN
{
    @ISA       = qw(Exporter);
    @EXPORT_OK = qw(NRPE_PACKET_VERSION_3
      NRPE_PACKET_VERSION_2
      NRPE_PACKET_VERSION_1
      NRPE_PACKET_QUERY
      NRPE_PACKET_RESPONSE
      MAX_PACKETBUFFER_LENGTH
      MAX_COMMAND_ARGUMENTS
      NRPE_HELLO_COMMAND
      DEFAULT_SOCKET_TIMEOUT
      DEFAULT_CONNECTION_TIMEOUT
      STATE_UNKNOWN
      STATE_CRITICAL
      STATE_WARNING
      STATE_OK);
}

use warnings;
use strict;

use Carp;
use Convert::Binary::C;
use Digest::CRC 'crc32';
use Nagios::NRPE::Utils qw(return_error);

use constant {

    # packet version identifier
    NRPE_PACKET_VERSION_3 => 3,
    NRPE_PACKET_VERSION_2 => 2,
    NRPE_PACKET_VERSION_1 => 1,

    # id code for queries and responses to queries
    NRPE_PACKET_QUERY    => 1,
    NRPE_PACKET_RESPONSE => 2,

    # max amount of data we'll send in one query/response
    MAX_PACKETBUFFER_LENGTH    => 1024,
    MAX_COMMAND_ARGUMENTS      => 16,
    NRPE_HELLO_COMMAND         => "_NRPE_CHECK",
    DEFAULT_SOCKET_TIMEOUT     => 10,
    DEFAULT_CONNECTION_TIMEOUT => 300,

    # /* service state return codes */
    STATE_UNKNOWN  => 3,
    STATE_CRITICAL => 2,
    STATE_WARNING  => 1,
    STATE_OK       => 0,
};

sub new
{
    my ($class, %options) = @_;
    my $self = {};

    bless $self, $class;
}

sub assemble
{
    my ($self, %options) = @_;

    # taken with modifications from common.h in nagios-nrpe
    my $c = Convert::Binary::C->new(ByteOrder => 'BigEndian', Alignment => 0);
    $self->{c} = $c;

    croak "ERROR: Cannot send Packet with empty buffer!"
      if (not defined $options{check});
    my $packed;
    if ($options{version} eq NRPE_PACKET_VERSION_2)
    {
        $packed = $self->assemble_v2(%options);
    }
    else
    {
        $packed = $self->assemble_v3(%options);
    }
    return $packed;

}

sub assemble_v3
{
    my ($self, %options) = @_;
    my $buffer = $options{check};
    my $len    = length($buffer);

    # In order for crc32 calculation to be correct we need to pad the buffer with \0
    # It seems that the buffer must be in multiples of 1024 so to achive this we use
    # some integer arithmetic to find the next multiple of 1024 that can hold our message
    my $packLen;
    {
        use integer;
        $packLen = (($len / 1024) * 1024) + 1024;
    }
    $buffer = pack("Z$packLen", $buffer);
    $len = length( $buffer) + 1;

    my $unpacked;
    $unpacked->{alignment}      = 0;
    $unpacked->{buffer_length}  = $len;
    $unpacked->{buffer}         = $buffer;
    $unpacked->{crc32_value}    = "\x00\x00\x00\x00";
    $unpacked->{packet_type}    = $options{type} // NRPE_PACKET_QUERY;
    $unpacked->{packet_version} = NRPE_PACKET_VERSION_3;
    $unpacked->{result_code}    = $options{result_code} // 2324;

    $self->{c}->parse(<<PACKET_STRUCT);
struct Packet{
  unsigned short   packet_version;
  unsigned short   packet_type;
  unsigned int     crc32_value;
  unsigned short   result_code;
  unsigned short   alignment;
  int              buffer_length;
  char             buffer[$len];
};
PACKET_STRUCT
    $self->{c}->tag('Packet.buffer', Format => 'String');

    my $packed = $self->{c}->pack('Packet', $unpacked);

    $unpacked->{crc32_value} = crc32($packed);
    $packed = $self->{c}->pack('Packet', $unpacked);
    return $packed;

}

sub assemble_v2
{

    my ($self, %options) = @_;
    my $unpacked = {};

    $unpacked->{buffer}         = $options{check};
    $unpacked->{crc32_value}    = "\x00\x00\x00\x00";
    $unpacked->{packet_type}    = $options{type} // NRPE_PACKET_QUERY;
    $unpacked->{packet_version} = NRPE_PACKET_VERSION_2;
    $unpacked->{result_code}    = $options{result_code} // 2324;

    $self->{c}->parse(<<PACKET_STRUCT);
struct Packet{
  unsigned short   packet_version;
  unsigned short   packet_type;
  unsigned int     crc32_value;
  unsigned short   result_code;
  char             buffer[1024];
};
PACKET_STRUCT
    $self->{c}->tag('Packet.buffer', Format => 'String');

    my $packed = $self->{c}->pack('Packet', $unpacked);

    $unpacked->{crc32_value} = crc32($packed);
    $packed = $self->{c}->pack('Packet', $unpacked);
    return $packed;

}

sub validate
{
    my ($self, $packet) = @_;
    my $unpacked = $self->disassemble($packet, 1);
    if (!$unpacked->{packet_version})
    {
        # If version is missing this is probably not an NRPE Packet.
        return undef;
    }
    my $checksum = $unpacked->{crc32_value};
    $unpacked->{crc32_value} = "\x00\x00\x00\x00";
    my $packed = $self->assemble(
                                 %{
                                     {
                                      check   => $unpacked->{buffer},
                                      version => $unpacked->{packet_version},
                                      type    => $unpacked->{packet_type},
                                      result_code => $unpacked->{result_code}
                                     }
                                  }
                                );
    if (crc32($packed) != $checksum)
    {
        return undef;
    }
    else
    {
        return 1;
    }
}

sub disassemble
{
    my ($self, $packet, $novalidate) = @_;
    if (!$packet)
    {
        return_error("Could not disassemble packet, it seems empty");
    }
    unless ($novalidate)
    {
        unless ($self->validate($packet))
        {
            return_error("Packet had invalid CRC32.");
        }
    }
    my $ver = unpack("n", $packet);
    my $unpacked = {};
    if ($ver)
    {
        if ($ver eq NRPE_PACKET_VERSION_2)
        {
            $unpacked = $self->disassemble_v2($packet);
        }
        else
        {
            $unpacked = $self->disassemble_v3($packet);
        }
    }
    else
    {
        return undef;
    }

    return $unpacked;
}

sub disassemble_v3
{
    my ($self, $packet) = @_;
    my @arr = unpack("n2 N n2 N Z*", $packet);
    my $unpacked = {};
    $unpacked->{packet_version} = $arr[0];
    $unpacked->{packet_type}    = $arr[1];
    $unpacked->{crc32_value}    = $arr[2];
    $unpacked->{result_code}    = $arr[3];
    $unpacked->{alignment}      = $arr[4];
    $unpacked->{buffer_length}  = $arr[5];
    $unpacked->{buffer}         = $arr[6];
    return $unpacked;
}

sub disassemble_v2
{
    my ($self, $packet) = @_;
    my @arr = unpack("n2 N n Z*", $packet);
    my $unpacked = {};
    $unpacked->{packet_version} = $arr[0];
    $unpacked->{packet_type}    = $arr[1];
    $unpacked->{crc32_value}    = $arr[2];
    $unpacked->{result_code}    = $arr[3];
    $unpacked->{buffer}         = $arr[4];
    return $unpacked;
}

sub packet_dump
{
    my $packet = shift;
    my $i;
    my $k;
    my $dump;
    my $l;
    my $ascii;
    my $c;
    for ($i = 0 ; $i < length($packet) ; $i += 16)
    {
        $l = $i + 16;
        if ($l > length($packet))
        {
            $l = length($packet);
        }
        $dump = sprintf("%04d - %04d: ", $i, $l);
        $ascii = "";
        for ($k = $i ; $k < $l ; $k++)
        {
            $c = (ord(substr($packet, $k, 1)));
            $dump .= sprintf("%02x ", $c);
            if (($c >= 32) && ($c <= 126))
            {
                $ascii .= chr($c);
            }
            else
            {
                $ascii .= ".";
            }
        }
        for ($k = 0 ; $k < ($i + 16 - $l) ; $k++)
        {
            $dump .= "   ";
        }
        print("packet_dump() " . $dump . " [" . $ascii . "]" . "\n");
    }
}

1;
